### ISSUE TITLE
Use Maven to build Copper.

### ISSUE THREAD

**Issue by schwerdf** [melt-umn/copper#21]
Copper currently relies on a hard-wired pathname to handle its dependency on ANT (for building its ANT task) and, for the same reason, cannot easily make use of JUnit or another unit-test framework.

I believe the dependencies can be better handled by using Maven instead of ANT to build Copper. As I understand the process, this would involve splitting the current codebase into five bundles: runtime, compile time, ANT task, test suite, and an "aggregate" bundle to bring everything together.

Switching to Maven will necessitate some changes to the automatic builds of Copper currently being run on the MELT system, but should not result in any substantial change to the final JARs.

This should be considered as blocking issues #11 and #12.

**Comment by tedinski** [comment]
If I understand things correctly, I think this could be just 3 bundles.

I believe the tests can live with the compile-time bundle, maven should have a separate "goal" for testing vs other things, as I understand it? As well as separate "compile time dependencies" that can bring in JUnit, for example? (I have yet to actually create a normal maven project from scratch, so I may be mistaken.)

And is there a reason to keep the ANT task separate from the compiler bundle? Is this just a dependency issue, bringing in ANT? I think Maven has optional dependencies to address this kind of odd case: https://maven.apache.org/guides/introduction/introduction-to-optional-and-excludes-dependencies.html Perhaps that will help avoid making it a separate bundle.

**Comment by schwerdf** [comment]
It is possible to use the `test` dependency scope to indicate that something like JUnit is not needed for the final build of a bundle, but in the Maven projects I have worked with, the tests are still given their own bundles, just to maintain a separation with the core source.

We have also found it necessary to give ANT tasks their own bundles in order to make the tasks available to run within Eclipse without explicitly providing the path to the containing JAR as is done now in `tests/build.xml`. Although this is not of immediate concern here, it could potentially be an issue if Copper is tracked in automatically as a Maven dependency.

**Comment by schwerdf** [comment]
I have pushed a Maven-built Copper to the `feature/MavenBuild` branch. The codebase is split roughly as planned above, although I further split the compile-time with the addition of a bundle for the legacy pipeline (Copper 0.5) that can be fully detached from the rest of Copper.

I would like someone else to test this before I merge it into `develop`. Expected behavior:

* Running `mvn test` from the `build/test` directory will run the new suite of regression tests. For now these just run Copper on every one of the test grammars.
* Running `mvn package` from the `build/release` directory will construct `CopperCompiler.jar` and `CopperRuntime.jar` in the `jar` directory.
* These JARs should be drop-in replacements for the ones currently in use, but I probably managed to break something during the refactoring, so the more Silver grammars they can be tried on, the better.

**Comment by tedinski** [comment]
Built and tested the jars with Silver & AbleC. No problems.

Regarding the Maven organization:

I ran a quick test that makes the following changes:

* Merges `build/release/pom.xml` `build/test/pom.xml` and `build/parent/pom.xml` into `pom.xml` (at the root)
  * This makes things a lot friendlier for new users. It's commonplace for there to be a `pom.xml` at the root, and that `mvn package` will build, test, and package the project's jars. This ensures users can get further without having to consult documentation.
  * You can still skip running tests with `mvn package -DskipTests`, if that was a motivating concern for separating these?
* Bug fix: noticed the `copper-test` module accidentally had the display name `copper-compiler`
* Another thing I noticed: we need to make it so JavaDoc gets generated again. Probably just a matter of configuring the right plugin?
* Moved a couple things, like sourceEncoding, to the parent instead of each child pom. I think there's more cleanup that can be done here, but I haven't looked too carefully yet because I wanted to see if you were interested.
* I was thinking about moving `build/aggregate` to just `aggregate` and `src/edu...runtime` to just `runtime` (ditto the rest), as this seems to mimic the style I see from other maven projects on github. Thoughts? Mostly, this just reduces the depth/length of paths.
* I'm still not totally certain why `test` is a separate module from `compiletime`, testing files are under `src/test` instead of `src/main` likewise `resources/test` instead of `resources/main`. I guess the objection is that's not separate enough?
* Possible change: instead of outputting `jar/CopperCompiler.jar` how about `target/CopperCompiler.jar`? This reduces the number of generated directories to just the expected ones (plus `mvn clean` will then delete it.)

Let me know what you think of this, and I can clean up and push my working copy as a separate branch to take a look at.

**Comment by schwerdf** [comment]
* Javadoc: I was waiting on this until we verified that this part worked right.
* Directory hierarchy: I would like to keep the `src` and `build` directories to maintain a separation between the code and anything else in the repository.
* Individual bundle names: We can ditch the `edu.umn.cs.melt` parts, but we should keep the `copper.` prefix (Eclipse projects named simply `compiletime` and `runtime` could be confusing).
* Tests: These are large regression tests rather than small-scale unit tests. Regression tests are kept separate in most of the Maven projects I work with because the test suites often take hours to run and thus are set up not to run unless explicitly invoked (and to enable distributing the source without the tests, although that is not an issue here).

Other than that, all these changes seem fine to me -- just push them on this same branch.

**Comment by tedinski** [comment]
* Javadoc: Ok. Just thought I'd mention it because I noticed `doc/javadoc` with an obsolete readme.
* Directory hierarchy: Ok about `src`. You're okay with `build` with only the `aggregate` directory in it?
* Names: Ahh, I hadn't thought about eclipse. I don't mind the names then.
* Tests: Fair enough. Except that my change putting `pom.xml` at the root means the tests do get run in a normal build, unless you `-DskipTests`, does that mean you object to having things merged at the root like I was suggesting?

**Comment by schwerdf** [comment]
`build` with only `aggregate` is good.
I am fine with putting `pom.xml` at the root, and I suppose that since the current test suite runs in seconds rather than hours, we can have it run automatically by default; but is there _no_ way to set up a separate build for the tests under your plan?

**Comment by tedinski** [comment]
I wouldn't say there's no way, I don't know. I'll take a little time to search around.

**Comment by tedinski** [comment]
Ah, this may jsut be a matter of switching from the "surefire" maven plugin, which does unit testing. To the "failsafe" maven plugin, which is evidently for integration testing. I'll investigate and try to get that working.

**Comment by tedinski** [comment]
Some more observations while I'm diving into this:

* We have a non-conventional location for resources: `resources/main` instead of `src/main/resources` (which puts it next to `src/main/java`). Is this intentional?
* The copper compiler says it also has Apache licensing, not just LGPL like all the other modules. Is this intentional?
* The `CopperCompiler.jar` also includes the runtime within itself. This is the same as the old behavior. I didn't realize we were doing that. I can add a `Class-Path` line to the manifest so it loads up the runtime jar in the same directory instead of duplicating this in the compiler jar, would this be a good change?

Good news though: I think I have the failsafe plugin successfully working, so it won't run tests with `mvn package` but will for `mvn verify`.

**Comment by schwerdf** [comment]
* I have no preference re: the resource location.
* The Copper compile-time includes some XML entity-handling code cribbed from Apache Commons, hence the need for the Apache License notice.
* It depends on how `CopperCompiler.jar` and `CopperRuntime.jar` are intended to be used. My idea was that projects using Copper through the API, like Silver, would switch to using the JARs of the individual Maven bundles (possibly tracked in as Maven dependencies), while the "release" JARs would be intended more for people experimenting with Copper on the command line. If this is the case, I would like to keep both the "release" JARs fully self-contained.

**Comment by tedinski** [comment]
I pushed my changes. It consists of the things we discussed above, except:

* I haven't renamed the modules, so they're all `src/edu...etc`.
* I left the resources in their current locations, rather than moving them to the conventional location.

These didn't seem that super important, and I decided I should stop working on this for now... :)

I also double checked that everything still works with Silver/AbleC.

Build with `mvn package`, run tests with `mvn verify`.
